libraries/apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/ApolloCall.kt

```diff
@@ -93,16 +106,36 @@ class ApolloCall<D : Operation.Data> internal constructor(
         .sendDocument(sendDocument)
         .enableAutoPersistedQueries(enableAutoPersistedQueries)
         .canBeBatched(canBeBatched)
+        .throwOnException(throwOnException)
         .build()
     return apolloClient.executeAsFlow(request)
   }
 
   /**
-   * A shorthand for `toFlow().single()`.
-   * Use this for queries and mutation to get a single [ApolloResponse] from the network or the cache.
-   * For subscriptions, you usually want to use [toFlow] instead to listen to all values.
+   * Retrieve a single [ApolloResponse] from this [ApolloCall], ignoring any cache misses or network errors.
+   *
+   * Use this for queries and mutations to get a single value from the network or the cache.
+   * For subscriptions or operations using `@defer`, you usually want to use [toFlow] instead to listen to all values.
+   *
+   * @throws ApolloException if the call returns zero or multiple valid GraphQL responses.
    */
   suspend fun execute(): ApolloResponse<D> {
-    return toFlow().single()
+    val responses = toFlow().toList()
+    val (errors, successes) = responses.partition { it.exception != null }
+    if (successes.size > 1) {
+      throw ApolloException("The operation returned multiple items, use .toFlow() instead of .execute()")
+    } else if (successes.isEmpty()) {
+      return if (errors.size == 1) {
+        errors[0]
+      } else if (errors.size > 1) {
+        errors[1].newBuilder()
+            .exception(ApolloCompositeException(errors[0].exception, errors[1].exception))
+            .build()
+      } else {
+        throw ApolloException("The operation did not emit any item, check your interceptor chain")
+      }
+    } else {
+      return successes.first()
+    }
   }
```

Comment:
Suggestion: use `when`, lift `return` outside of the branches and sort the `when` cases by number of items

```suggestion
    return when(successes.size) {
      0 -> {
        when(errors.size) {
          0 -> throw ApolloException("The operation did not emit any item, check your interceptor chain")
          1 -> errors.first()
          else -> {
            errors[1].newBuilder()
                .exception(ApolloCompositeException(errors[0].exception, errors[1].exception))
                .build()
          }
        }
      }
      1 -> {
        successes.first()
      }
      else -> {
        throw ApolloException("The operation returned multiple items, use .toFlow() instead of .execute()")
      }
    }
```

Reply:
Also might be worth adding an `ApolloCompositeException` that takes > 2 suppressed exceptions to account for this case?